SwiftCompilerSources/Sources/Optimizer/DataStructures/InstructionRange.swift

```diff
@@ -181,3 +181,54 @@ struct InstructionRange : CustomStringConvertible, NoReflectionChildren {
     blockRange.deinitialize()
   }
 }
+
+extension InstructionRange {
+  enum PathOverlap {
+    case containsPath
+    case containsBegin
+    case containsEnd
+    case overlappedByPath
+    case disjoint
+  }
+
+  /// Return true if any exclusive path from 'begin' to 'end' includes an instruction in this exclusive range.
+  ///
+  /// Precondition: 'begin' dominates 'end'.
+  func overlapsWithPath(pathBegin: Instruction, pathEnd: Instruction, _ context: some Context) -> PathOverlap {
+    assert(pathBegin != pathEnd, "expect an exclusive path")
+    if contains(pathBegin) {
+      // Note: pathEnd != self.begin here since self.contains(pathBegin)
+      if contains(pathEnd) { return .containsPath }
+      return .containsBegin
+    }
+    if contains(pathEnd) {
+      return .containsEnd
+    }
+    // Neither end-point is contained. If a backward path walk encouters this range, then it must overlap this
+    // range. Otherwise, it is disjoint.
+    var backwardBlocks = BasicBlockWorklist(context)
+    defer { backwardBlocks.deinitialize() }
+    backwardBlocks.pushIfNotVisited(pathEnd.parentBlock)
+    while let block = backwardBlocks.pop() {
+      if !blockRange.inclusiveRangeContains(block) {
+        backwardBlocks.pushIfNotVisited(contentsOf: block.predecessors)
+        continue
+      }
+      if block == pathBegin.parentBlock {
+        let endInBlock = block == pathEnd.parentBlock ? pathEnd : block.terminator
+        for inst in ReverseInstructionList(first: endInBlock) {
+          if contains(inst) {
+            return .overlappedByPath
+          }
+          if inst == pathBegin {
+            return .disjoint
+          }
+        }
+        // fallthru: pathBegin was not reached
```

Comment:
Can this fall-through really happen? Shouldn't the loop always find the `pathBegin` instruction?
If so, maybe change this to a `fatalError`

Reply:
restructured and commented to make it clear how this case can happen.